convert realism to functions, add idle animation

vnajack opened the pull request:
- renamed file from mouthCueProcessor to animationProcessor
- converted realism (blinking, poses) to functions for readability
- implements code for idle animation (will probably need adjustments later)

NOTE: Before merging into dev, we need to add the actual frames of the avatars with appropriate file names, so we can convert this to a draft PR until then

- related to #169 
- related to #204
- closes #209

vnajack committed:
convert realism to functions, add idle animation

- rename file from mouthCueProcessor to animationProcessor
- convert realism (blinking, poses) to functions
- implement beginning code of idle animation

LoganParker commented on speaking-portal-project/src/animationFactory/animationProcessor.ts:
```diff
@@ -0,0 +1,155 @@
+        // TODO: determine if avatar will do a body pose or be in idle animation
+
+        // Idling or Posing
+        if(idle){
+            breathPhase = calculateBreathPhase(currentSec, lastPoseChange, breathPhase)
+            frameData.concat(translateBreathFrame(breathPhase))
+        } else {
+            breathPhase = 0
+        }
+
```
I think we will eventually need to add additional checks here, for example if the character is not idle, what if it's raising its hand, or doing a little dance, etc?

vnajack replied:
Oh definitely, I forgot to add the todo here! Good catch

LoganParker commented on speaking-portal-project/src/animationFactory/animationProcessor.ts:
```diff
@@ -0,0 +1,155 @@
+    return frameData
+}
+
+function willBlink(currentSec: number, frameDur: string, lastBlink: number) {
```
Great refactor here! Thank you, I was in the process of doing this so I really appreciate you doing it!!

vnajack replied:
Thank YOU! I was struggling coding this morning, so I'm glad we discussed it!

LoganParker commented on speaking-portal-project/src/animationFactory/animationProcessor.ts:
```diff
@@ -0,0 +1,155 @@
+    return result
+}
+
+// TODO: if we make breathing a type, we can probably get rid of this function
```
rather than creating an individual type for breathing, we might want to create a type for the character? That way it can hold states for breathing/blinking/motion of other kinds?

vnajack replied:
Ed and I were discussing it, and there's a lot to unpack so we're checking to see if types are appropriate or if we should define a class, or what it is we want to do. This will probably be a part of the refactoring we do over time...

LoganParker commented on speaking-portal-project/src/animationFactory/animationProcessor.ts:
```diff
@@ -0,0 +1,155 @@
+        console.log("t=%d, frameDur: %d, lastBlink: %d, idling? %s, lastPoseChange: %d",
+                    currentSec, frameDur, lastBlink, lastPoseChange, idle, lastPoseChange)
+
+        // Start building frame's path
```
We should eventually extract this into a separate function, but not needed for now!

vnajack replied:
Yeah, I think making it a separate function will need to wait until we have all the realism pieces together.. unless you have an idea for how to implement it now!

LoganParker commented on speaking-portal-project/src/animationFactory/animationProcessor.ts:
```diff
@@ -0,0 +1,155 @@
+        // Blinking
+        blink = willBlink(currentSec, frameDur, lastBlink)
+        if (blink) {
+            lastBlink += 1
```
Is lastBlink just holding a variable relating to how many times it blinks? or should we be making lastBlink = the frame it last blinked?

vnajack replied:
Last blink is the second in which the character last blinked. It is used so that the avatar doesn't blink in the same second.

If you see double blinks, it's because the avatar probably blinked at the end of a second and then again at the beginning of another second :)

LoganParker commented on speaking-portal-project/src/animationFactory/animationProcessor.ts:
```diff
@@ -0,0 +1,155 @@
+}
+
+function calculateBreathPhase(currentSec: number, lastPoseChange: number, breathPhase: number) {
+    /**
+     * Determines this frame's breath phase value based on time and most recent pose
+     *
+     * @remarks
+     * - A normal respiratory rate in healthy adults is roughly 12 to 20 breaths per minute
+     * - Our avatars' breathing cycle will take 4 seconds, which results in ~15 breaths/min
+     * - Breathing Phases:
+     *      0 = rest
+     *      1 = inhaling or "breathing"
+     *      2 = lungs full
+     *      3 = exhaling or "breathing"
+     *
+     * @returns the new value of breathPhase
+     *
+     **/
```
A task we might want to setup for next week is creating some documentation including things like these comments - we could even just move the comments from the code into their own separate documentation document?

vnajack replied:
Yeah, it's a lot of info for a function definition, but I wanted to keep in mind what I was doing while writing the code. We can definitely move stuff around!!

LoganParker commented on speaking-portal-project/src/animationFactory/animationProcessor.ts:
```diff
@@ -0,0 +1,155 @@
+     * @returns A string to indicate the idle breathing frame
+     *
+     **/
+    if (breathPhase == 0 ){
+        return "_rest" // rest, arms relaxed and down
+    } else if (breathPhase == 1 || breathPhase == 3){
+        return "_breathing" // inhale and exhale (transition between rest and peak)
+    } else { // == 2
+        return "_lungsFull" // lungs are full, arms are out
```
This will be dependent on the # of frames we have correct?

I have a sneaking suspicion that if we only have 3 frames, it might look either too fast or too jittery, but once I export the assets I suppose we'll have a better idea - they do look somewhat jittery in photoshop though with just 3 frames

vnajack replied:
No! It's not dependent on the frames! It actually does follow time in seconds! I think it should work okay, but I would like to test it out

vnajack commented on speaking-portal-project/src/animationFactory/animationProcessor.ts:
```diff
@@ -0,0 +1,155 @@
+            breathPhase = calculateBreathPhase(currentSec, lastPoseChange, breathPhase)
+            frameData.concat(translateBreathFrame(breathPhase))
+        } else {
+            breathPhase = 0
```
```suggestion
            breathPhase = 0
            // TODO: add poses
```

vnajack committed:
Add comment for TODO reminder

LoganParker committed:
Update err handling?

LoganParker committed:
Merge branch 'development' into realism-breathing

LoganParker committed:
Merge branch 'realism-breathing' of https://github.com/speaking-portal-project-team-a/The-Speaking-Portal-Project into realism-breathing

LoganParker committed:
Merge branch 'development' into realism-breathing

LoganParker committed:
Update characterSelect var to avatar

vnajack committed:
update to realism animation processor

- had to run npm install again, so `package-lock.json` is updated
- removed unnecessary avatar frames
- renamed files for animation processor and readability
- updated and fixed some small bugs in `animationProcessor.ts`
- updated `characterSelect` to `avatar`
- commented out the idle animation code, as it's not ready yet but it's a start

vnajack committed:
Merge branch 'realism-breathing' of https://github.com/speaking-portal-project-team-a/The-Speaking-Portal-Project into realism-breathing

vnajack committed:
update again woops

LoganParker reviewed (approved):
Thanks Veronica! These changes look incredible, im stoked to get our assets looking like real things!

edouarde1 reviewed (approved):
This looks great Veronica. I know I have not been doing much for the realism this week, so if you and Logan and feeling good with it, it's got my stamp.

I know we still have a lot to discuss about the typings/ Data representations of the added components for realism. Looking forward figuring that out!
